### PR TITLE
[Datahub] Add an API form generator for OGC API Records

### PR DESCRIPTION
**Pull request by cmoinier**
This PR creates an API form component to handle customization of OGC API urls.

It also adds a date-range picker component to gn-ui (because such an input was in the original mockup, and will be used in gn-editor in the future).

**Funded by Métropole de Lille**

image

### CODE REVIEW

File: apps/datahub/src/app/record/custom-api/custom-api.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/custom-api/custom-api.component.html:1]: IMO
- the should be a shared feature component
- it's not custom-api, for now it's OGC API Features

File: apps/datahub/src/app/record/custom-api/custom-api.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/custom-api/custom-api.component.html:3]: Why a height ? Is it supposed to be scrollable ?
Otherwise the height is built from the content.

BTW, you should use `class="h-[512px]"` instead

File: apps/datahub/src/app/record/custom-api/custom-api.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/custom-api/custom-api.component.html:36]: Does not seem optimzed to switch from 23 to 20 like this.
BTW, do we have a zeplin or figma ?

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/record-metadata/record-metadata.component.html:124]: Note that in the long run, there will be several kinds of APIs forms. For instance
- OGC API Features
- WFS
- IGN download service
- IGN extraction service

**Review by fgravin — COMMENTED** [review]
Hi @cmoinier, not a real review but some thoughts from here to there :relaxed:

File: apps/datahub/src/app/record/custom-api/custom-api.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:1]: Okay for the shared feature component point, but I named it like that in prevision of the future, to avoid having to rename it. Since there will be numerous kinds of APIs forms, doesn't it make sense to have a generic name?

File: apps/datahub/src/app/record/custom-api/custom-api.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:3]: This is a WIP, the style is not done at all, I barely did the structure here ^^. This `div` was copied from the `data view` part of the record page (which yes, has a height for some reason), just to have a `div` that would go across the screen. I'm planning on doing styling after everything else.

File: apps/datahub/src/app/record/custom-api/custom-api.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:36]: There is a figma, I got the link to it after committing everything I've done here. I will use it to have better styling values from now on. Styling is not done/approximative right now, I'm focusing on the functionalities first.

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/record-metadata/record-metadata.component.html:124]: Noted. This output is here to allow the custom API cards to open the panel to customize the URL, which isn't available for pre-defined API cards.

File: apps/datahub/src/app/record/custom-api/custom-api.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:1]: How about : `api-custom-url` ? Any other idea?

Comment (fgravin) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:1]: will it handle all kinds of APIs ? eg OGC, WFS, STAC and so on ? Or is it just for this concern (OGC API Features) ?

Comment (cmoinier) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:1]: Yes, this component will be used in the future for all kinds of APIs. For now, only OGC though

Comment (fgravin) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:1]: `record-api-form` ?

Comment (cmoinier) [reply apps/datahub/src/app/record/custom-api/custom-api.component.html:1]: Sounds good!

File: apps/datahub/src/app/record/record-apis/record-apis.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/record-apis/record-apis.component.html:16]: translation keys must be normalized.

File: libs/ui/elements/src/lib/record-api-form/record-api-form.component.ts
Comment (fgravin) [inline libs/ui/elements/src/lib/record-api-form/record-api-form.component.ts:22]: Why do you use Observables everywhere ?
Classic change detection is meant for this.
BTW, we could take the opportunity to test signals !

File: apps/datahub/src/app/record/record-apis/record-apis.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/record-apis/record-apis.component.html:43]: Too much duplicated code.
Probably a way to handle that with template outlet + context.
Anyway, don't jump into that yet cause I don't see the point of those 2 tabs, wait Lucile to give feedbacks.

File: apps/datahub/src/app/record/record-apis/record-apis.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/record-apis/record-apis.component.html:65]: The name `isCustom` for the event is confusing.

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/record-metadata/record-metadata.component.html:136]: You should take the habit to not write french text in the template, it's a source for missing some translation :upside_down_face:

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/record-metadata/record-metadata.component.html:124]: wondering if it could go in mdview state .. any thought @tkohr @jahow ?

**Review by fgravin — COMMENTED** [review]
Thanks @cmoinier for your progress on this.
I've made some general remarks regarding code practice.
Let me know if it make sense for you :relaxed:

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/record-metadata/record-metadata.component.html:136]: Translations are not done yet :slightly_smiling_face:

File: apps/datahub/src/app/record/record-apis/record-apis.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/record-apis/record-apis.component.html:65]: Good catch, we reused this output for something else and forgot the change the name

File: apps/datahub/src/app/record/record-apis/record-apis.component.html
Comment (cmoinier) [reply apps/datahub/src/app/record/record-apis/record-apis.component.html:16]: Translations are not done yet (I was updating the todo list when you reviewed :slightly_smiling_face: )

File: libs/ui/elements/src/lib/record-api-form/record-api-form.component.ts
Comment (cmoinier) [reply libs/ui/elements/src/lib/record-api-form/record-api-form.component.ts:22]: Actually, the other solutions we tried to get the different inputs to react to changes made from the code (eg : reset to the initial values), were not working with classic change detection, hence the Observables. But let me know if you have other suggestions, happy to hear them.

Comment (fgravin) [reply libs/ui/elements/src/lib/record-api-form/record-api-form.component.ts:22]: Did you try `ChangeDetectorRef` ?
Actually I don't know what is the best practice about that.

Comment (cmoinier) [reply libs/ui/elements/src/lib/record-api-form/record-api-form.component.ts:22]: I believe we did, right @tkohr ?

Comment (tkohr) [reply libs/ui/elements/src/lib/record-api-form/record-api-form.component.ts:22]: Indeed, we had some change detection problems when switching cards which we were not able to resolve (even by manually calling `ChangeDetectorRef.detectChanges()`)

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (tkohr) [reply apps/datahub/src/app/record/record-metadata/record-metadata.component.html:124]: I think I'd rather keep it in the `record-metadata` "state" like all the `displaySomething$` observables. But maybe you see some advantage in the facade for later needs?

**Comment by fgravin** [comment]
Could we see a screenshot ?? :relaxed:

File: apps/datahub-e2e/src/e2e/home.cy.ts
Comment (tkohr) [inline apps/datahub-e2e/src/e2e/home.cy.ts:81]: Don't forget to remove the `only` here.

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (tkohr) [inline apps/datahub/src/app/record/record-metadata/record-metadata.component.html:148]: I think it'd be nice to move this html to the form component.

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.ts
Comment (tkohr) [inline apps/datahub/src/app/record/record-metadata/record-metadata.component.ts:28]: Do we still need this `Output`?

File: libs/ui/elements/src/lib/api-card/api-card.component.ts
Comment (tkohr) [inline libs/ui/elements/src/lib/api-card/api-card.component.ts:28]: This could probably also go into a setter, but both works I guess.

File: libs/ui/elements/src/lib/api-card/api-card.component.ts
Comment (tkohr) [inline libs/ui/elements/src/lib/api-card/api-card.component.ts:37]: We could maybe still name the method a little more explicitly (eg. `openRecordApiFormPanel` even if it's long?)

File: libs/ui/elements/src/lib/api-card/api-card.component.ts
Comment (tkohr) [inline libs/ui/elements/src/lib/api-card/api-card.component.ts:39]: I don't think we have a test for this ternary open/close logic, not sure...

**Review by tkohr — COMMENTED** [review]
Great work @cmoinier ! Thanks for your continued effort on this. I've left some minor comments. Besides that, I still see two points to address:

- Currently, the form breaks on mobile. I guess we can just hide the three dots button on mobile or replace it with the copy url button. Not sure if this has been discussed.
- Comparing the form with the mockup I still see quite some differences, like font sizes, paddings and margins, line breaks of the labels.

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (fgravin) [inline apps/datahub/src/app/record/record-metadata/record-metadata.component.html:130]: I think that this section should be within the `datahub-record-apis` component, it's purely related.
WSYT @tkohr ?

**Review by fgravin — COMMENTED** [review]
Very good work, you've added a bunch of e2e & unit tests so it's very complete :clap: 

Just yes like @tkohr says, there are many slight differences between the mockup and your design (fonts, sizes, margin etc..). Please try to replicate the mockup with accuracy, it's all about details :smile: 

Thanks

File: apps/datahub/src/app/record/record-metadata/record-metadata.component.html
Comment (tkohr) [reply apps/datahub/src/app/record/record-metadata/record-metadata.component.html:130]: Ah right, I missed that too. Ideally, I'd see the lines I commented below in the `record-api-form.component` and then the `record-api-form.component` go into the `record-metadata.comonent`. WDYT @cmoinier @fgravin ?

**Comment by cmoinier** [comment]
Thanks @tkohr @fgravin for your reviews. I have corrected most of the things you mentionned as per the last commit. I still need to improve the design before it's considered done. :slightly_smiling_face:

**Comment by fgravin** [comment]
> Thanks @tkohr @fgravin for your reviews. I have corrected most of the things you mentionned as per the last commit. I still need to improve the design before it's considered done. 🙂

Thanks @cmoinier we will have a look.
BTW I am wondering how do you test ? I think that having a metadata which references an OGC API Features resources in the `support-services` compo would help.
I don't know how hard it is though.

Thanks

**Comment by cmoinier** [comment]
@fgravin I added a dump with a metadata that has two OGC API Features resources.

**Review by fgravin — DISMISSED** [review]
Still some issues with the rendering.
You don't handle the width of your form correctly, see the "preview" container which behave the same.
- on full screen, the form is too large
- on mobile, it's too narrow.

Also, the style of the copy text input is not like the mockup.

**Comment by fgravin** [comment]
On my tests, the cross of the close button is bigger than the text.
image

**Comment by fgravin** [comment]
Using the support-services locally, the dataset "Accroche velo" with the ogc api features has lost the "similar records" panel.
Is it related to the PR ?

**Comment by cmoinier** [comment]
Noted for the rendering issues, I'm on it. 

For the "similar records" panel, yes, it is related to the PR. I think I have accidentally deleted the similar record link and therefore, the panel isn't displaying anymore. I will add it again and push another dump.

**Review by tkohr — APPROVED** [review]
Thanks, @cmoinier, everything looks good on my side! I just added a last commit to add some opacity to the transition and remove some unnecessary code, I still came across.

I let you have a look at it and wait for the CI to pass, before merging :-)

**Pull request merged**
